feat: add support for "down" media queries in "MediaQuery.is()" #11122

ncoden opened the pull request:
<!--- --------------------------------------------------------------------- -->
<!---                 Please fill the following template                    -->
<!---             Your pull request may be ignored otherwise                -->
<!--- --------------------------------------------------------------------- -->

## Description
<!--- Describe your changes in detail. Include any relevant information     -->
<!--- (reasons, difficulties, links to web references, related issues...)   -->

This pull request add support for the "down" keyword in `MediaQuery.is()` thanks to the new method `MediaQuery.upTo()`. It also add `MediaQuery.only()` and unit tests for `MediaQuery.atLeast()`, `MediaQuery.only()` and `MediaQuery.upTo()`.

<!--- Bugs and new features/improvements must be presented and discussed in -->
<!--- an issue first. Please create one if there is no issue related to     -->
<!--- this pull request.                                                    -->
- Closes https://github.com/zurb/foundation-sites/issues/11122

### Unit tests
The implemented process to test media queries is the following
* **Start a local server on port**
  Local ressources with `file://` cannot be used in an iframe in Chrome. We serve the Foundation repositoryon port `3042`.
* **Dynamically create an iframe.**
  Create an iframe Element, then inject into it the pre-compiled mediaQuery CSS configuration and jQuery/Foundation javascript from the local server.
* **Run tests.**
  Initialize the iframe version of MediaQuery, resize the iframe and run tests.

### Changes
* Validate `MediaQuery.is()` input and returns an error for invalid modifiers
* Add `MediaQuery.upTo()` and `MediaQuery.only()`
* Add support of `down` breakpoint modifier to "MediaQuery.is()`
* Add unit tests for `MediaQuery.atLeat()`, `MediaQuery.only()` and `MediaQuery.upTo()`
* Fix an issue when MediaQuery is initialized twice.
* Improve and update MediaQuery documentation

## Types of changes
<!--- What types of changes does your code introduce?                       -->
<!--- Put an `x` in all the boxes that apply:                               -->
- [ ] Documentation
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to change)

## Checklist (all required):
<!--- Go over all the following points, and put an `x` in the boxes.        -->
<!--- If you're unsure about any of these, don't hesitate to ask.           -->
- [x] I have read and follow the [CONTRIBUTING](CONTRIBUTING.md) document.
- [x] There are no other pull request similar to this one.
- [x] The pull request title is descriptive.
- [x] The template is fully and correctly filled.
- [x] The pull request targets the right branch (`develop` or `develop-v...`).
- [x] My commits are correctly titled and contain all relevant information.
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly to my changes (if relevant).
- [x] I have added tests to cover my changes (if relevant).
- [x] All new and existing tests passed.

<!--- --------------------------------------------------------------------- -->
<!---       For more information, see the CONTRIBUTING.md document          -->
<!---         Thank you for your pull request and happy coding ;)           -->
<!--- --------------------------------------------------------------------- -->

ncoden committed:
fix: validate "MediaQuery.is()" input and returns an error for invalid modifiers

ncoden committed:
feat: add "MediaQuery.upTo()" and support of "down" mq to "MediaQuery.is()"

ncoden committed:
test: add base for MediaQuery unit tests

ncoden committed:
test: prevent cache in iframe in MediaQuery unit tests

ncoden committed:
refactor: move MediaQuery query name resolution to "_getQueryName()"

ncoden committed:
fix: fix MediaQuery.upTo()

MediaQuery.upTo() should return whether the screen is within a breakpoint or smaller. The previous implementation had an incorrect algorithm.

The new implementation is more divided and documented, making the algorithm way more clear and auditable.

ncoden committed:
test: add tests for MediaQuery.atLeast()

ncoden committed:
test: add tests for MediaQuery.upTo()

ncoden committed:
feat: add "MediaQuery.only()"

Add method:
* `MediaQuery.only(size)`
  Checks if the screen is within the given breakpoint.
  If smaller than the breakpoint of wider than its upper limit, returns false.

ncoden committed:
test: add tests for "MediaQuery.only()"

ncoden committed:
docs: improve MediaQuery JS documentation & add only/down modifiers

ncoden commented:
@DanielRuf @SassNinja for review 👋

DanielRuf commented on docs/pages/media-queries.md:
```diff
@@ -197,22 +197,24 @@ Get the name of the current breakpoint with `MediaQuery.current`.
+Foundation.MediaQuery.is('medium only');
+Foundation.MediaQuery.only('medium');
+
+// ↓ True for "medium" or smaller
```
small and wide are two different things. It is either smaller and bigger or narrower and wider.
But not smaller and wider.

ncoden replied:
What about `smaller` and `larger`? This is how the documentation already describe breakpoint.

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
 
+  /**
+   * Checks if the screen is within the given breakpoint.
+   * If smaller than the breakpoint of wider than its upper limit, returns false.
```
See my other note about wider / bigger.

DanielRuf replied:
, => it

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
+   * Checks if the screen is within a breakpoint or smaller.
+   * @function
+   * @param {String} size - Name of the breakpoint to check.
+   * @returns {Boolean} `true` if the breakpoint matches, `false` if it's wider.
```
wider / bigger

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
+  upTo(size) {
+    const nextSize = this.next(size);
+
+    // If the next breakpoint do not match, the screen is smaller than
```
do => does

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
+    const nextSize = this.next(size);
+
+    // If the next breakpoint do not match, the screen is smaller than
+    // the upper limit of this breakpoint.
```
the => The

ncoden replied:
No.
> the screen is smaller than (...) the upper limit of this breakpoint.

I break comments at 80 when possible.

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
+    if (bpModifier === 'down') {
+      return this.atLeast(bpSize);
+    }
+    // At leat the breakpoint  (included)
```
point  (included) = point (included)   (one space less)

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -121,20 +123,64 @@ var MediaQuery = {
+
+    throw new Error(`
+      invalid breakpoint passed to MediaQuery.is().
+      Expected a breakpoint breakpoint name formatted like "<size> <modifier>", got "${size}".
```
breakpoint breakpoint => breakpoint (duplication)

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -154,6 +200,43 @@ var MediaQuery = {
+    if (queryIndex === -1) {
+      throw new Error(`
+        unknown breakpoint "${size}" passed to MediaQuery.next().
+        Ensure it is present in your Sass settings "$breakpoints".
```
settings "$breakpoints" => "$breakpoints" setting

DanielRuf commented on js/foundation.util.mediaQuery.js:
```diff
@@ -154,6 +200,43 @@ var MediaQuery = {
+    if (typeof value === 'object')
+      return value.name;
+    throw new TypeError(`
+      invalid value passed to MediaQuery._getQueryName().
```
invalid => Invalid

DanielRuf commented on test/javascript/lib/utils.js:
```diff
@@ -40,4 +40,29 @@
+
+  /**
+  * Inject a script with the attributes of the `opts` object into the
+  * given `context` HTML element. The created script is garanted to be executed
```
garanted => guaranteed

DanielRuf commented on test/javascript/util/mediaQuery.js:
```diff
@@ -0,0 +1,148 @@
+      plugin.only('large').should.be.false;
+    });
+
+    it('returns "false" when wider than given breakpoint', function () {
```
than => than the

ncoden committed:
docs: use "smaller" and "larger" to describe breakpoints in Grid/Media Query doc

ncoden committed:
docs: fix various typo and import errors wording in MediaQuery

ncoden commented:
Fixed, and resolved conflicts with `develop`

andycochran commented:
@DanielRuf Do you approve the additional commits @ncoden made after your review? 

Might you be able to resolve the conflicts? Looks like most of them could be resolved by trashing the lock files and reinstalling dependencies.

DanielRuf commented:
I'll try to find some time to answer in the next few days.

DanielRuf reviewed (approved):
LGTM

DanielRuf committed:
chore: update lockfiles

joeworkman committed:
update http-server version

This fixes travis testing

joeworkman committed:
Fix Windows tests

Remove the http-server start since this is not supported on windows as a background process.

joeworkman committed:
Use Mocha assets instead of http-server

Stopped using http-server

joeworkman committed:
Stop using http-server

joeworkman commented:
This pull request has been mentioned on **Foundation Open Source Community**. There might be relevant details there:

https://foundation.discourse.group/t/foundation-for-sites-v6-6-0-is-here-farout/30/1
